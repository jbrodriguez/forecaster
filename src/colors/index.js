// @flow

import adapter from '../lib/adapter'

const palette = {
  primary: '#878D93', // baseText
  secondary: '#B9BCC0', // altText
  accent: '#ffffff', // accentText

  background: '#0F1B27',
  bgTone1: '#2F3B46',
  bgTone2: '#1A2733',

  contrast: '#6D6345',

  divider: '#1E2934',
}

const colors = adapter(palette)

export default colors
