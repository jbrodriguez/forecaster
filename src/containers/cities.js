// @flow

import React, { PureComponent } from 'react'
import { View, TouchableOpacity, Text, RefreshControl, Platform } from 'react-native'

import { bindActionCreators } from 'redux'
import { connect, Dispatch } from 'react-redux'
import { NavigationActions } from 'react-navigation'
import { SwipeListView } from 'react-native-swipe-list-view'

import type { TCity, TGui, ASetRefreshing, ASetCurrentCity, ADeleteCity, ARefreshAll } from '../typings'
import { getGui, isRefreshing } from '../modules/env'
import { citiesByOrder, setCurrentCity, deleteCity, refreshAll } from '../modules/model'

import HeaderRight from './cities.header'
import CityProxy from '../components/cityProxy'

type Props = {
  cities: TCity[],
  gui: TGui,
  refreshing: boolean,
  setRefreshing: (payload: boolean) => ASetRefreshing,
  setCurrentCity: (payload: number) => ASetCurrentCity,
  deleteCity: (payload: number) => ADeleteCity,
  refreshAll: () => ARefreshAll,
  dispatch: Dispatch,
}

class Cities extends PureComponent<Props> {
  static navigationOptions = {
    title: 'Cities',
    headerRight: <HeaderRight />,
  }

  componentDidMount() {
    // this.props.refresh()
  }

  onRefresh = () => {
    this.props.refreshAll()
  }

  onDelete = (id: number, row) => () => {
    row.closeRow()
    this.props.deleteCity(id)
  }

  goTo = (id: number) => () => {
    const { cities } = this.props

    const city = cities.find(item => item.id === id)

    // this shouldn't happen, since we're selecting an item from the cities array
    if (!city) {
      return
    }

    this.props.setCurrentCity(id)
    this.props.dispatch(NavigationActions.navigate({ routeName: 'City', params: { title: city.name } }))
  }

  renderRow = data => <CityProxy gui={this.props.gui} city={data.item} goTo={this.goTo} />

  renderHidden = (data, rowMap) => {
    const { gui: { s, c } } = this.props
    const row = rowMap[data.item.id.toString()]
    return (
      <View style={[s.flx_i, s.flx_row, s.aic, s.jcsb, c.g_background]}>
        <TouchableOpacity
          style={[
            s.aic,
            s.jcc,
            {
              position: 'absolute',
              top: 0,
              bottom: 0,
              left: 0,
              width: 75,
            },
            c.g_alert,
          ]}
          onPress={this.onDelete(data.item.id, row)}
        >
          <Text style={[s.f5, c.c_white]}>Delete</Text>
        </TouchableOpacity>
      </View>
    )
  }

  render() {
    const { cities, gui: { s, c }, refreshing } = this.props

    const bgStyle = Platform.OS === 'ios' ? c.g_transparent : null

    if (cities.length === 0) {
      return (
        <View style={[s.flx_i, s.pt4, c.g_background]}>
          <Text style={[s.f4, s.tc, c.c_secondary]}>Add a city using the plus icon on the top right</Text>
        </View>
      )
    }

    return (
      <View style={[s.flx_i, c.g_background]}>
        <SwipeListView
          useFlatList
          data={cities}
          keyExtractor={item => item.id.toString()}
          renderItem={this.renderRow}
          renderHiddenItem={this.renderHidden}
          refreshControl={
            <RefreshControl
              style={bgStyle}
              refreshing={refreshing}
              onRefresh={this.onRefresh}
              tintColor={c.accent}
              title="Refreshing ..."
              titleColor={c.accent}
              colors={[c.accent]}
              progressBackgroundColor={c.primary}
            />
          }
          recalculateHiddenLayout
          leftOpenValue={75}
          disableLeftSwipe
        />
      </View>
    )
  }
}

const mapStateToProps = state => ({
  cities: citiesByOrder(state),
  gui: getGui(state),
  refreshing: isRefreshing(state),
})
const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(
    {
      setCurrentCity,
      deleteCity,
      refreshAll,
      // setRefreshing,
    },
    dispatch,
  ),
  dispatch,
})

export default connect(mapStateToProps, mapDispatchToProps)(Cities)
