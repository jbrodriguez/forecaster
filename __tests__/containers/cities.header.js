import { StyleSheet } from 'react-native'
import React from 'react'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

import Tachyons from 'react-native-style-tachyons'

import colors from '../../src/colors'
import { HeaderRightBase } from '../../src/containers/cities.header'

describe('CitiesHeader', () => {
  Tachyons.build({ rem: 16 }, StyleSheet)

  const gui = { s: Tachyons.styles, c: colors, z: Tachyons.sizes }
  const search = jest.fn()
  const settings = jest.fn()

  test('renders correctly - nohero', () => {
    const tree = renderer.create(<HeaderRightBase gui={gui} search={search} settings={settings} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
