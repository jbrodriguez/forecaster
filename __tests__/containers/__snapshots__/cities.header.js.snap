// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesHeader renders correctly - nohero 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      Object {
        "alignItems": "center",
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontSize": 36,
          },
          Object {
            "marginRight": 8,
          },
          Object {
            "color": "#B9BCC0",
          },
        ]
      }
    >
      +
    </Text>
  </View>
  <View
    style={
      Object {
        "marginRight": 8,
      }
    }
  />
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontSize": 36,
          },
          Object {
            "marginRight": 8,
          },
          Object {
            "color": "#B9BCC0",
          },
        ]
      }
    >
      s
    </Text>
  </View>
</View>
`;
